src/test/java/org/walkmod/javalang/compiler/SymbolVisitorAdapterTest.java

```diff
@@ -1095,18 +1095,61 @@ public void testSingletonList() throws Exception {
 	}
 
 	@Test
-	public void testTemplateResults() throws Exception {
+	public void testGenericResultForParameterizedTypeWithInterfaceWitness() throws Exception {
+		itestTypeParameterWithWitness("<C extends Comparable<?>>",
+				"A<C>",
+				"Comparable<java.io.Serializable>",
+				"A<Comparable<java.io.Serializable>>",
+				"A<java.lang.Comparable<java.io.Serializable>>");
+	}
+
+	@Test
+	public void testGenericResultForParameterizedTypeWithClassWitness() throws Exception {
+		itestTypeParameterWithWitness("<C extends Comparable<?>>",
+				"A<C>",
+				"Integer",
+				"A<Integer>",
+				"A<java.lang.Integer>");
+	}
+
+	@Test
+	public void testGenericResultForTypeVariableWithInterfaceWitness() throws Exception {
+		itestTypeParameterWithWitness("<C>",
+				"C",
+				"Comparable<java.io.Serializable>",
+				"Comparable<java.io.Serializable>",
+				"java.lang.Comparable<java.io.Serializable>");
+	}
+
+	@Test
+	public void testGenericResultForTypeVariableWithClassWitness() throws Exception {
+		itestTypeParameterWithWitness("<C>",
+				"C",
+				"Integer",
+				"Integer",
+				// root for improvement
+				"java.lang.Integer");
+	}
+
+	private void itestTypeParameterWithWitness(String typeParam, final String returnType, final String witnessType, final String effectiveType, String expectedType) throws Exception {
```

Comment:
I also prefer test* for the test, but the i does not sound descriptive to me. What about `checkTypeParameterWithWitness`?